dask/async.py

```diff
@@ -567,3 +572,139 @@ def sortkey(item):
     ('tuple', ('x', 1))
     """
     return (type(item).__name__, item)
+
+
+def get_deps(dsk):
+    """ Get dependencies and dependents from dask dask graph
+
+    >>> dsk = {'a': 1, 'b': (inc, 'a'), 'c': (inc, 'b')}
+    >>> dependencies, dependents = get_deps(dsk)
+    >>> dependencies
+    {'a': set([]), 'c': set(['b']), 'b': set(['a'])}
+    >>> dependents
+    {'a': set(['b']), 'c': set([]), 'b': set(['c'])}
+    """
+    dependencies = dict((k, get_dependencies(dsk, k)) for k in dsk)
+    dependents = reverse_dict(dependencies)
+    return dependencies, dependents
+
+
+def ndependents(dependencies, dependents):
+    """ Number of total data elements that depend on key
+
+    >>> dsk = {'a': 1, 'b': (inc, 'a'), 'c': (inc, 'b')}
+    >>> dependencies, dependents = get_deps(dsk)
+
+    >>> sorted(ndependents(dependencies, dependents).items())
+    [('a', 3), ('b', 2), ('c', 1)]
+    """
+    result = dict()
+
+    roots = [k for k, v in dependents.items() if not v]
+
+    result.update(dict((r, 1) for r in roots))
+
+    leaves = [k for k, v in dependencies.items() if not v]
+
+    for leaf in leaves:
+        _ndependents(leaf, result, dependencies, dependents)
+
+    return result
+
+
+def _ndependents(key, result, dependencies, dependents):
+    """ Helper function for ndependents """
+    if key not in result:
+        deps = dependents[key]
+        result[key] = sum(
+            [_ndependents(k, result, dependencies, dependents)
+             for k in deps]) + 1
+    return result[key]
+
+
+def child_max(dependencies, dependents, scores):
+    """
+
+    The maximum of all child scores minus a half.
+    This half score lets us pass a little bit of information
+    down the graph about dependence.  Otherwise things tend to flatten out.
+
+    >>> dsk = {'a': 1, 'b': 2, 'c': (inc, 'a'), 'd': (add, 'b', 'c')}
+    >>> scores = {'a': 3, 'b': 2, 'c': 2, 'd': 1}
+    >>> dependencies, dependents = get_deps(dsk)
+
+    >>> sorted(child_max(dependencies, dependents, scores).items())
+    [('a', 3), ('b', 2), ('c', 2.5), ('d', 2.0)]
+    """
+    result = dict()
+
+    leaves = [k for k, v in dependencies.items() if not v]
+
+    for leaf in leaves:
+        result[leaf] = scores[leaf]
+
+    roots = [k for k, v in dependents.items() if not v]
+
+    for root in roots:
+        _child_max(root, scores, result, dependencies, dependents)
+
+    return result
+
+
+def _child_max(key, scores, result, dependencies, dependents):
+    """ Recursive helper function for child_max """
+    if key not in result:
+        deps = dependencies[key]
+        result[key] = max([_child_max(k, scores, result, dependencies,
+                                      dependents) for k in deps]) - 0.5
+    return result[key]
+
+
+def dfs(dependencies, dependents, key=lambda x: x):
+    """
+
+    >>> dsk = {'a': 1, 'b': 2, 'c': (inc, 'a'), 'd': (add, 'b', 'c')}
```

Comment:
This is a sign that I wanted a doctest but couldn't yet think of a title line for the docstring itself.  More a peculiarity of my doctest-heavy programming process.

I've added a title line and general prose both here and above.